File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Scenarios/MonC1DispatchFailureScenario.cs
// Verifies MON-C1 claim progress and observer isolation behavior.
using RuntimeMonitoring.Client.Support;
using RuntimeMonitoring.Shared;
using Zlink.HttpClient;

namespace RuntimeMonitoring.Client.Scenarios;

internal static class MonC1DispatchFailureScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        using var serviceA = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(35))
            .Build();
        using var serviceB = ZLinkHttpClient.Create(options.ServiceBUrl)
            .Timeout(TimeSpan.FromSeconds(35))
            .Build();

        await serviceA.Post(
                $"/runtime/observer/start/{RuntimeMonitoringNames.SpotChannel}")
            .Async<ObserverIsolationStatusRes>();
        await serviceA.Post("/admin/application-gate/reset").Async<object>();
        await serviceB.Post("/admin/spot-weight/exclude").Async<object>();

        var evidenceBaseline =
            (await serviceA.Get("/evidence").Async<string[]>()).Body.Length;
        var blockedRequest = serviceA.Post("/spot/profile/request")
            .Body(new ProfileReq("application-gate", "mon-c1-gated"))
            .Timeout(TimeSpan.FromSeconds(30))
            .Async<ProfileRes>();
        await serviceA.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(
                ["application-gate-enter|rid=svc-a|marker=mon-c1-gated"],
                [],
                TimeoutMilliseconds: 3000,
                AfterIndex: evidenceBaseline))
            .Async<string[]>();

        var before = (await serviceA.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.SpotChannel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;
        ZlinkStreamAssert.Ensure(
            before.IsReady && before.Placement.IsAvailable,
            "MON-C1 RouteMesh readiness regressed while an application handler was blocked.");
        ZlinkStreamAssert.Ensure(
            !blockedRequest.IsCompleted,
            "MON-C1 application gate did not hold the request handler.");

        await serviceB.Post("/admin/spot-weight/include").Async<object>();
        await serviceA.Post("/admin/spot-weight/exclude").Async<object>();
        var terminalReply = (await serviceB.Post("/spot/profile/request")
            .Body(new ProfileReq("parallel", "mon-c1-terminal"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(
            terminalReply.ProviderRid == "svc-b"
            && terminalReply.Marker == "mon-c1-terminal",
            "MON-C1 separate request did not complete while the application gate was held.");
        ZlinkStreamAssert.Ensure(
            !blockedRequest.IsCompleted,
            "MON-C1 application gate was released before the separate request completed.");

        for (var cycle = 0; cycle < 8; cycle++)
        {
            await serviceB.Post("/admin/spot-weight/exclude").Async<object>();
            _ = await serviceA.Get(
                    $"/runtime/snapshot/{RuntimeMonitoringNames.SpotChannel}")
                .Async<MeshRuntimeSnapshotRes>();
            await Task.Delay(125);
            await serviceB.Post("/admin/spot-weight/include").Async<object>();
            _ = await serviceA.Get(
                    $"/runtime/snapshot/{RuntimeMonitoringNames.SpotChannel}")
                .Async<MeshRuntimeSnapshotRes>();
            await Task.Delay(125);
        }

        var active = (await serviceA.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.SpotChannel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;
        ZlinkStreamAssert.Ensure(
            active.IsReady && active.Sequence >= before.Sequence,
            "MON-C1 runtime status did not progress with the application gate held.");

        var beforeRelease = await WaitForObserverAsync(
            serviceA,
            status => status.NormalEventCount >= 2,
            "MON-C1 normal observer did not progress.");
        ZlinkStreamAssert.Ensure(
            !beforeRelease.SlowConsumerReleased
            && beforeRelease.NormalLatestSequence > 0,
            "MON-C1 slow observer blocked the normal observer.");

        await serviceA.Post("/runtime/observer/release")
            .Async<ObserverIsolationStatusRes>();
        var afterRelease = await WaitForObserverAsync(
            serviceA,
            status => status.SlowConsumerFailed
                      && status.SlowLatestSequence > 0,
            "MON-C1 slow consumer failure was not isolated.");
        ZlinkStreamAssert.Ensure(
            afterRelease.NormalLatestSequence
            >= beforeRelease.NormalLatestSequence,
            "MON-C1 normal observer regressed after slow consumer failure.");

        var resynced = (await serviceA.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.SpotChannel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;
        ZlinkStreamAssert.Ensure(
            resynced.Sequence >= afterRelease.SlowLatestSequence
            && resynced.Channels.Any(channel =>
                channel.ChannelName == RuntimeMonitoringNames.SpotChannel),
            "MON-C1 snapshot resync did not retain the latest channel state.");

        await serviceA.Post("/admin/application-gate/release").Async<object>();
        var gatedReply = (await blockedRequest).Body;
        ZlinkStreamAssert.Ensure(
            gatedReply.ProviderRid == "svc-a"
            && gatedReply.Marker == "mon-c1-gated",
            "MON-C1 gated request did not complete after release.");

        var followUp = (await serviceB.Post("/spot/profile/request")
            .Body(new ProfileReq("recovery", "mon-c1-recovery"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(
            followUp.Value == "profile:recovery",
            "MON-C1 messaging stopped after the observer consumer failure.");

        await serviceA.Post("/admin/spot-weight/include").Async<object>();
        Console.WriteLine("scenario MON-C1 passed");
    }

    private static async Task<ObserverIsolationStatusRes> WaitForObserverAsync(
        ZLinkHttpClient service,
        Func<ObserverIsolationStatusRes, bool> predicate,
        string failureMessage)
    {
        for (var attempt = 0; attempt < 100; attempt++)
        {
            var status = (await service.Get("/runtime/observer/status")
                .Async<ObserverIsolationStatusRes>()).Body;
            if (predicate(status))
                return status;
            await Task.Delay(100);
        }
        throw new InvalidOperationException(failureMessage);
    }
}

File: framework/languages/node/e2e/RuntimeMonitoring/Client/Scenarios/mon-c1-dispatch-failure-scenario.ts
// MON-C1: 느리거나 실패한 observer가 다른 작업을 막지 않는다 시나리오를 검증한다.
import type { ProfileReq, ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { type ManagedProcess, startServiceB, waitForPortState } from '../Support/managed-service';
import {
  readRouteStatus,
  routeStatusesFromEvidence,
  waitForRouteStatus,
  delay
} from '../Support/public-status';
import { ensure } from '../Support/scenario-assert';

export async function runMonC1(options: ClientOptions): Promise<ManagedProcess> {
  const before = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.isReady && status.peers.some((peer) => peer.nodeRid === 'svc-b'),
    'MON-C1 initial RouteMesh status did not contain a ready svc-b peer.'
  );

  await postJson(options.serviceUrl, '/observer/slow/start', {});
  await postJson(options.serviceUrl, '/observer/failing/start', {});
  await postJson<string[]>(options.serviceUrl, '/evidence/wait', {
    containsAll: ['observer-slow|sequence=', 'observer-failing|sequence=', 'observer-failed|message=monitoring observer failure for e2e'],
    containsAnyGroups: [],
    timeoutMilliseconds: 10000
  });
  const slowSequence = readSlowObserverSequence(await getJson<readonly string[]>(options.serviceUrl, '/evidence'));

  await postJson<object>(options.serviceBUrl, '/shutdown', {});
  await waitForPortState(options.serviceBUrl, false, 'MON-C1 expected svc-b to stop while the slow observer was blocked.');
  const removed = await waitForRouteStatus(
    options.serviceUrl,
    (status) => !status.peers.some((peer) => peer.nodeRid === 'svc-b'),
    'MON-C1 normal observer did not expose the stopped svc-b peer.'
  );

  const restarted = startServiceB(options, 'svc-b-mon-c1');
  try {
    await waitForPortState(options.serviceBUrl, true, 'MON-C1 expected svc-b to restart.');
    const restored = await waitForRouteStatus(
      options.serviceUrl,
      (status) => status.isReady
        && status.peers.some((peer) => peer.nodeRid === 'svc-b' && peer.state === 1)
        && status.channels.some((channel) => channel.channelName === 'monitor.profile' && channel.isReady)
        && BigInt(status.sequence) > BigInt(removed.sequence),
      'MON-C1 normal observer did not expose the restored svc-b peer.'
    );

    const request: ProfileReq = { value: 'observer', marker: `mon-c1-${Date.now()}` };
    const reply = await postJson<ProfileRes>(options.triggerUrl, '/profile/request/service-b', request);
    ensure(reply.value === 'profile:observer' && reply.providerRid === 'svc-b', 'MON-C1 request did not complete while observers were blocked.');
    await postJson<string[]>(options.serviceBUrl, '/evidence/wait', {
      containsAll: [`profile-request|rid=svc-b|marker=${request.marker}|value=observer`],
      containsAnyGroups: [],
      timeoutMilliseconds: 10000
    });

    const status = await readRouteStatus(options.serviceUrl);
    ensure(status.isReady, 'MON-C1 normal RouteMesh status was not available after observer failure.');
    ensure(BigInt(status.sequence) >= BigInt(restored.sequence), 'MON-C1 current status moved backwards after recovery.');
    ensure(BigInt(status.sequence) > BigInt(slowSequence), 'MON-C1 status did not advance while the slow observer was blocked.');
    ensure(BigInt(restored.sequence) > BigInt(before.sequence), 'MON-C1 RouteMesh status did not record the peer transition.');

    await waitForObservedRoute(options.serviceUrl, (observed) =>
      observed.some((candidate) => !candidate.isReady
        && candidate.peers.some((peer) => peer.nodeRid === 'svc-b' && peer.state !== 1))
      && observed.some((candidate) => candidate.peers.some((peer) => peer.nodeRid === 'svc-b'))
      && observed.some((candidate) => BigInt(candidate.sequence) >= BigInt(restored.sequence))
    );

    await postJson(options.serviceUrl, '/observer/slow/release', {});
    console.log('scenario MON-C1 passed');
    return restarted;
  } catch (error) {
    await restarted.stop();
    throw error;
  }
}

function readSlowObserverSequence(lines: readonly string[]): string {
  const line = lines.find((entry) => entry.startsWith('observer-slow|sequence='));
  if (line === undefined) throw new Error('MON-C1 slow observer did not receive its first status callback.');
  const sequence = line.slice('observer-slow|sequence='.length);
  if (!/^\d+$/.test(sequence)) throw new Error('MON-C1 slow observer returned an invalid sequence.');
  return sequence;
}

async function waitForObservedRoute(
  url: string,
  predicate: (statuses: ReturnType<typeof routeStatusesFromEvidence>) => boolean
): Promise<void> {
  const deadline = Date.now() + 15_000;
  while (Date.now() < deadline) {
    const statuses = routeStatusesFromEvidence(await getJson<readonly string[]>(url, '/evidence'));
    if (predicate(statuses)) return;
    await delay();
  }
  throw new Error('MON-C1 normal observer did not provide removal, restoration, and current status evidence.');
}
